### ISSUE TITLE
reduce_scatter produces inaccurate results when input dtype is bf16

### ISSUE THREAD

**Issue by sshonTT** [tenstorrent/tt-metal#36749]
### Component / Area

kernels

### Issue Type (optional)

Bad Outputs

### Observed

In the tt-mlir runtime, the all_reduce operation is decomposed into a combination of reduce_scatter and all_gather. 
We've observed that when input tensors contain values near the limits of the BF16 range, reduce_scatter yields poor results, leading to significant model accuracy degradation.

Implementing all_reduce via all_gather followed by a local reduce works correctly without precision loss.

Converting the input dtype to FP32 before running reduce_scatter and apply this [PR #36579](https://github.com/tenstorrent/tt-metal/pull/36579) also resolves the issue.

There appears to be a precision or range-handling issue within the reduce_scatter op specifically when the input is BF16. While we could use dtype conversion as a workaround, it introduces unnecessary overhead. Please investigate and fix the underlying issue in the reduce_scatter kernel.

### Expected

reduce_scatter op yields valid output when using bf16 input.

### 1. Steps (exact commands)

Using branch this (https://github.com/tenstorrent/tt-metal/tree/sshon/bf16-reduce-scatter-accuracy),

run `python -m pytest tests/ttnn/unit_tests/operations/ccl/test_reduce_scatter_bf16_precision.py -svv`

In this script, there is an result when using all_gather + local reduce which pcc is 1.0

### 2. Input data / link or description

Input from test script would be sufficient.

### 3. Frequency

Always

### 1. Software Versions

Ubuntu 22.04, latest metal commit

### 2. Hardware Details

Warmhole, llmbox (t3k)

### Is this a regression?

Unknown

### Priority

P0

### Impact

It blocks running gpt oss model from frontend because of accuracy issue.

**Comment by skhatriTT** [comment]
@fahmedchanTT can you please triage this issue?

**Comment by zzigler-tt** [comment]
Changing this to P1 and will have the team triage next week. 
@fahmedchanTT @cmaryanTT

**Comment by cmaryanTT** [comment]
If it's just one of our vanilla reducitons, I'm fairly sure those all run fp32 accumulation by default. This was part of a accurate-by-default push for Borys' area over the last few months. @bbradelTT can you confirm

**Comment by AleksKnezevic** [comment]
yeah, this case def seems to have some numerics issues and as stated above, fp32 accum does help. 

For llama, enabling f32 accumulation with reduce scatter did not resolve the PCC issue. Note, I also [fixed](https://github.com/tenstorrent/tt-metal/compare/main...aknezevic/rs_accum) the preset `max_dest_size`

**Comment by smeydanshahiTT** [comment]
When looking a bit closely at the values themselves rather than the PCC, I found that the Max ATOL Delta is 768.0. This is important as during the reduction, there are 512 sized gaps. When we add a new value, the hardware performs an addition, and the result must be rounded to the nearest representable BF16 value. However, since the gap is 512, the hardware can easily round up/down incorrectly. The delta being 786 is important as 512 * 1.5 = 786, indicating that over the course of the reduce_scatter the value drifted by 1.5 steps due to the rounding errors.

I do believe that switching to fp32 dtypes and accum would resolve the issue, but will continue looking into it.

**Comment by bbradelTT** [comment]
> If it's just one of our vanilla reducitons, I'm fairly sure those all run fp32 accumulation by default. This was part of a accurate-by-default push for Borys' area over the last few months. [@bbradelTT](https://github.com/bbradelTT) can you confirm

Yes, generic reductions should use fp32 accumulation by default.

**Comment by cmaryanTT** [comment]
Yup, confirmed our normal reductions (not RS), use fp32 accumulation by default. Also, the test case from @sshonTT does AG plus a torch.sum (not ttnn.sum), which is fp32 or something compensated internally.

**Comment by cmaryanTT** [comment]
I think we should wrap this up as follows:
1. Quick fix for the Forge team is to switch to fp32 dtype
2. @smeydanshahiTT can merge the fp32 accumulate into RS, and that should be the preferred solution. We'll perf test it, and perhaps turn it on by default if it's not too much overhead. Merging this will take a bit, because of the testing, existing model impact and Sam's other priorities.
3. @vladimirjovanovicTT will provide a repro for the llama variant. If there's something there, it may be a different problem - but I'm fairly confident in 1 & 2 above.

I'm knocking this down to a P1, since 1 is the pragmatic workaround for the short term. We'll still try to get 2 in fairly soon.

**Comment by AleksKnezevic** [comment]
Correct me if I'm wrong, but don't we need fp32 accum enabled even if we switch to fp32 inputs? @sshonTT had a change to enable fp32 accum for fp32 inputs [here](https://github.com/tenstorrent/tt-metal/pull/36579)
